uniffi/src/ffi/rustcalls.rs

```diff
@@ -0,0 +1,251 @@
+///   "out parameter" which will be updated with any error that occurred during the function's
+///   execution.
+/// - After the call, if code is `CALL_ERROR` then `error_buf` will be updated to contain
+///   the serialized error object.   The consumer is responsible for freeing `error_buf`.
+///
+/// ## Layout/fields
+///
+/// The layout of this struct is important since consumers on the other side of the FFI need to
+/// construct it.  If this were a C struct, it would look like:
+///
+/// ```c,no_run
+/// struct RustCallStatus {
+///     int8_t code;
+///     RustBuffer error_buf;
+/// };
+/// ```
+///
+/// #### The `code` field.
+///
+///  - `CALL_SUCESS` (0) for successful calls
+///  - `CALL_ERROR` (1) for calls that returned an `Err` value
+///  - `CALL_PANIC` (2) for calls that panicked
+///
+/// #### The `error_buf` field.
+///
+/// - For `CALL_ERROR` this is a `RustBuffer` with the serialized error.  The consumer code is
+///   responsible for freeing this `RustBuffer`.
+#[repr(C)]
+pub struct RustCallStatus {
+    pub code: i8,
+    // code is signed because unsigned types are experimental in Kotlin
+    pub error_buf: MaybeUninit<RustBuffer>,
+    // error_buf is MaybeUninit to avoid dropping the value that the consumer code sends in:
+    //   - Consumers should send in a zeroed out RustBuffer.  In this case dropping is a no-op and
+    //     avoiding the drop is a small optimization.
+    //   - If consumers pass in invalid data, then we should avoid trying to drop in.  In
+    //     particular, we don't want to try to free any data the consumer has allocated.
+}
+
+#[allow(dead_code)]
+const CALL_SUCCESS: i8 = 0; // CALL_SUCCESS is set by the calling code
+const CALL_ERROR: i8 = 1;
+const CALL_PANIC: i8 = 2;
+
+// A trait for errors that can be thrown to the FFI code
+//
+// This gets implemented in uniffi_bindgen/src/scaffolding/templates/ErrorTemplate.rs
+pub trait FfiError: RustBufferViaFfi {}
+
+// Generalized rust call handling function
+// callback should map errors to a RustBuffer that contains them
+fn make_call<F, R>(out_status: &mut RustCallStatus, callback: F) -> R::Value
+where
+    F: panic::UnwindSafe + FnOnce() -> Result<R, RustBuffer>,
+    R: IntoFfi,
+{
+    let result = panic::catch_unwind(|| {
+        // We should call:
+        //
+        // init_panic_handling_once();
+        //
+        // This is called in ffi-support::call_with_result_impl().  The current plan is to make it
+        // `pub` in ffi-support and call it from here.
+        callback().map(|v| v.into_ffi_value())
+    });
+    match result {
+        // Happy path.  Note: no need to update out_status in this case because the calling code
+        // initializes it to CALL_SUCCESS
+        Ok(Ok(v)) => v,
+        // Callback returned an Err.
+        Ok(Err(buf)) => {
+            out_status.code = CALL_ERROR;
+            unsafe {
+                out_status.error_buf.as_mut_ptr().write(buf);
```

Comment:
It would be useful to briefly summarize the safety invariants here. IIUC it's specifically the `write` that's unsafe, and we're upholding the invariant that it's a valid and aligned pointer, by assumption that the passed `out_status` pointer is valid.

Reply:
I think the main reason it's unsafe is it's `MaybeUninit`, so we won't be dropping any previously initialized data.  I'll add few comments about that one, it's definitely not clear from the code.

I didn't think about alignment issues, I guess we just have to assume that all the bindings code gives us a properly aligned struct.